File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] a, int target) {
      int m = a.length, n = a[0].length;
      int st = 0, end = n * m - 1;

      while(st <= end) {
        int mid = st + (end - st) / 2;
        int midElt = a[mid/n][mid%n];

        if(midElt == target) return true;

        if(target < midElt) {
            end = mid - 1;
        } else {
            st = mid + 1;
        }
      }
      return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int r = matrix.size();
        int c = matrix[0].size();
        
        int st = 0;
        int end = r * c - 1;
        
        while (st <= end) {
            int mid = st + (end - st) / 2;
            int midEle = matrix[mid/c][mid%c];
            
            if (midEle == target) return true;
            else if (target < midEle) end = mid - 1;
            else st = mid + 1;
        }
        
        return false;
    }
};